[display] as a html snippet

stonier opened the pull request:
Raw output comparisons (on the console):

![comparisons](https://user-images.githubusercontent.com/687378/56441461-2a530100-62bb-11e9-9e64-517c37768cfb.png)

Embedded in a html document:

![html_snippet](https://user-images.githubusercontent.com/687378/56441433-042d6100-62bb-11e9-9839-52605d6c5dc8.png)

stonier committed:
[display] html trees

stonier committed:
redundant

stonier committed:
docs

stonier commented:
Minor bugfix in the original code...and I'm sure there are better ways to do string formatting, but I'm not inclined to go down a rabbit hole for that yet.

I did refactor things a little to reduce the amount of repeated code that could create duplicate bugs.

sachit-wp commented on py_trees/display.py:
```diff
@@ -183,6 +212,33 @@ def generate_lines(root, internal_indent):
+
+    .. seealso:: :meth:`py_trees.display.ascii_tree`
+    """
+    lines = ascii_tree(root, show_status, visited, previously_visited, indent, symbols=html_symbols)
```
Given the simplicity, maybe just handle this in ascii_tree?

stonier replied:
Originally had them bundled but started reflecting on the user perspective. They're going to find it easier to think of `ascii_tree` and `xhtml_tree` without having to concern themselves with symbols, formatters, ... It will also be more robust for users in the future - the symbol formatter under the hood is quite likely to change, but `xhtml_tree` signature should be stable.

Having said that, I've only dragged myself halfway there. I really ought to just break it finally and have:

* private `_generate_text_tree` method that does the grunt work
* public `ascii_art_tree` method
* public `unicode_art_tree` method that falls back to `ascii_art_tree` when it discovers unicode is not available
* public `xhtml_art_tree` method

sachit-wp commented on tests/test_display.py:
```diff
@@ -23,24 +23,65 @@ def test_symbols():
+def test_html_tree():
+    console.banner("Ascii/Html Snapshots - Comparison Check")
+
+    def print_tree(snapshot_visitor, tree):
```
Probably can try parsing the data as xml instead of just printing it.

stonier replied:
Good idea, now parses the xml and it helped ensure it's now a valid xhtml snippet, not just html (problems with the non-breaking space character)

stonier committed:
now producing valid xhtml (not merely html)

stonier committed:
html_tree->xhtml_tree and changelog update

sachit-wp commented:
Pulled the latest changes and wrote this script:
```
import py_trees
a = py_trees.behaviours.Success()
b = py_trees.behaviours.Success()
c = c = py_trees.composites.Sequence(children=[a, b])
c.tick_once()
py_trees.display.xhtml_tree(c)
```

and my output was as follows:
```<code>\n<text>[-]</text> Sequence [<br/>\n<text>&#xa0;</text><text>&#xa0;</text><text>&#xa0;</text><text>&#xa0;</text><text>--></text> Success [<br/>\n<text>&#xa0;</text><text>&#xa0;</text><text>&#xa0;</text><text>&#xa0;</text><b><text>--></text></b> <b>Success [</b><br/>\n</code>```

This is what it looks like in JSFiddle:
https://jsfiddle.net/396pnw4x/

Seems like there's something wrong. 

Maybe add a regexp as part of the test to make sure that there's the name of the behaviour as well as the status as part of the text as well?

stonier committed:
kill a stray bracket

stonier committed:
more tests

stonier commented:
Aye, good catch. Stray bracket found and more tests added.

sachit-wp commented on tests/test_display.py:
```diff
@@ -14,6 +14,22 @@
+##############################################################################
+
+
+def assert_banner():
```
BTW probably more meaningful if these methods were prefixed with print_ to indicate their purpose.

stonier replied:
I've got the same calls across the other `py_trees` repositories. I'd have to change them all to be consistent. Would prefer to leave as is as it's a nice-to-have only.

stonier replied:
Done.

sachit-wp commented on py_trees/display.py:
```diff
@@ -183,6 +211,33 @@ def generate_lines(root, internal_indent):
+    .. seealso:: :meth:`py_trees.display.ascii_tree`
+    """
+    lines = ascii_tree(root, show_status, visited, previously_visited, indent, symbols=html_symbols)
+    lines = lines.replace("\n", "<br/>\n")
```
Only need the <br/> here. The \n is not really doing anything useful.

stonier replied:
I know it doesn't usefully effect the xhtml, but it does logically break it up a little so it's easier to read rather than one giant text blob if you happen to be trying to debug the actual xhtml under the hood.

sachit-wp replied:
True, but then the xhtml output has these stray \n lines everywhere and the user has to manually replace them after calling this API in order to view the data properly in a browser.

[From Example Above] See here:
https://jsfiddle.net/396pnw4x/show

sachit-wp replied:
I was being silly. Sorry about that. Decided to copy paste __repr__ instead of the output value

sachit-wp commented on tests/test_display.py:
```diff
@@ -23,24 +40,81 @@ def test_symbols():
+    # Visited
+    snippets["visited_ascii_tree"] = py_trees.display.ascii_tree(
+            tree.root,
+            visited=snapshot_visitor.visited,
```
BTW test using show_status=True as well would be nice.

stonier replied:
Added

stonier committed:
display show status tests, consistent tree method frontends

stonier committed:
changelog

stonier commented:
Added tests for `show_status` and responded to the other comment. 

Also refactored `ascii_tree`/`xhtml_tree` into `ascii_tree`/`unicode_tree`/`xhtml_tree` as frontends to the backend `_generate_text_tree` as discussed. This will make them more robust against implementation changes and also avoids the fallback that subverts unicode to ascii behaviour based on a rather arbitrary lookup on stdout's encoding.

There is also a `dot_graph` -> `dot_tree` refactor to be consistent with the text methods.

Not planning on any more changes unless there is an iteration required on the above.

stonier committed:
assert_ -> print_assert_

sachit-wp commented on py_trees/display.py:
```diff
@@ -123,74 +230,109 @@ def post_tick_handler(snapshot_visitor, behaviour_tree):
+
+    Examples:
+
+    .. code-block:: python
```
:)
